stdlib/join/merge_join.go

```diff
@@ -0,0 +1,522 @@
+		}
+	}
+	return labels
+}
+
+// Inserts `rows` into s.products, while maintaining sort order. Returns a position and a bool.
+// If the bool == true, that means it's safe to join all the items in s.products up to and including
+// the returned position.
+//
+// Returns true under 2 circumstances:
+//   (1) Inserting `rows` completes the left and right pair for a given product
+//   (2) `rows` was inserted at an index greater than 0, and all of the products that
+//        come before it only have entries on the opposite side.
+//
+// If condition 1 is true, we can join everything up to and including the index where
+// `rows` was inserted.
+//
+// If condition 2 is true and condition 1 is false, we can join up to, but not including,
+// the index where rows was inserted.
+func (s *joinState) insert(key *joinKey, rows joinRows, isLeft bool) (int, bool) {
+	if len(s.products) == 0 {
+		p := newJoinProduct(key, rows, isLeft)
+		s.products = []joinProduct{p}
+		return 0, false
+	}
+
+	found := false
+	position := 0
+	for i, product := range s.products {
+		if product.key.equal(*key) {
+			if isLeft {
+				product.left = rows
+			} else {
+				product.right = rows
+			}
+			s.products[i] = product
+			found = true
+			position = i
+			break
+		} else if key.less(product.key) {
+			newProducts := make([]joinProduct, 0, len(s.products)+1)
+			newProducts = append(newProducts, s.products[:i]...)
+			newProducts = append(newProducts, newJoinProduct(key, rows, isLeft))
+			newProducts = append(newProducts, s.products[i:]...)
+			s.products = newProducts
+			found = true
+			position = i
+			break
+		}
+		position = i
+	}
+
+	if !found {
+		s.products = append(s.products, newJoinProduct(key, rows, isLeft))
+		position = len(s.products) - 1
+	}
+
+	if s.products[position].isDone() {
+		return position, true
+	}
+
+	// The only condition where we would not emit a signal to join and flush
+	// the list of products is if we received a bunch of `joinRows` all
+	// from the same side.
+	//
+	// So if the previous product is only populated on one side, then it is either the first
+	// item in the list, or all of the items before it are also populated on the same side.
+	canJoin := false
+	if position > 0 {
+		position--
+		prev := s.products[position]
+		if isLeft {
+			canJoin = prev.left.len() == 0 && prev.right.len() > 0
+		} else {
+			canJoin = prev.left.len() > 0 && prev.right.len() == 0
+		}
+	}
+	return position, canJoin
+}
+
+func (s *joinState) join(ctx context.Context, method string, fn *JoinFn, joinable int, mem memory.Allocator) ([]table.Chunk, error) {
+	if fn.compiled == nil {
+		err := fn.Prepare(s.left.schema, s.right.schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	joined := make([]table.Chunk, 0, joinable+1)
+	for i := 0; i <= joinable; i++ {
+		prod := s.products[i]
+		p, ok, err := prod.evaluate(ctx, method, *fn, mem)
+		if err != nil {
+			return joined, err
+		}
+		if !ok {
+			continue
+		}
+		joined = append(joined, *p)
```

Comment:
I wonder if `evaluate` should produce multiple chunks? It's possible for join to produce many rows if there are many duplicate join keys on both sides---you end up with `ln * rn` rows. Maybe file an issue for it?

Reply:
I just pushed up a commit that limits each of the output chunks to 1000 rows each. Take a look and let me know if that's what you had in mind.